src/sql/src/parsers/cursor_parser.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use snafu::{ensure, ResultExt};
+use sqlparser::keywords::Keyword;
+use sqlparser::tokenizer::Token;
+
+use crate::error::{self, Result};
+use crate::parser::ParserContext;
+use crate::statements::cursor::{CloseCursor, DeclareCursor, FetchCursor};
+use crate::statements::statement::Statement;
+
+impl ParserContext<'_> {
+    pub(crate) fn parse_declare_cursor(&mut self) -> Result<Statement> {
+        let _ = self.parser.expect_keyword(Keyword::DECLARE);
+        let cursor_name = self
+            .parser
+            .parse_object_name(false)
+            .context(error::SyntaxSnafu)?;
+        let _ = self
+            .parser
+            .expect_keywords(&[Keyword::CURSOR, Keyword::FOR]);
+
+        let mut is_select = false;
+        if let Token::Word(w) = self.parser.peek_token().token {
+            if w.keyword == Keyword::SELECT {
```

Comment:
What about [`WITH`](https://docs.greptime.cn/en/reference/sql/with), it's select too.